### PR TITLE
KIN-356 Product filters visual components functionality

### PR DESCRIPTION
**Pull request by CarlosMay7**
[KIN-356](https://loopcrack.atlassian.net/browse/KIN-356) Extra Large [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=loopcrack-org&utm_campaign=badge_info)

### Problem

The customer needs a way to find products in an easy and fast way

### Solution

Create different kinds of filters on the category page 

### Requirements checklist

- [x] It must have a functional search bar
- [x] The search bar must display all the products related to the user's input
- [x] The cards displayed from the search bar must have an image and information related to the product
- [x] When you try to search for some product using the search bar it must change the URL (Only pressing enter, when clicking on a displayed card will be implemented on the backend integration but it also changes the URL)
- [x] The display selection buttons must change the product catalog display between the grid and list
- [x] It must have a sorting button
- [x] It must present an option to change how many products will be shown
- [x] It must have a pagination

### Testing

- [x] Manually tested all functionalities

## Demo

[Category Demo](https://drive.google.com/file/d/1haZYWbA7OvASrjx4zrCXd8gpGiBnEqkv/view?usp=sharing)

[KIN-356]: https://loopcrack.atlassian.net/browse/KIN-356?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/Config/Routes.php [hunk 1]
```diff
@@ -73,6 +74,9 @@
 $routes->post('password_reset/(:any)', [CtrlLogin::class, 'passwordReset']);
 $routes->post('password_set/(:any)', [CtrlLogin::class, 'passwordSet']);
 $routes->post('logout', [CtrlLogin::class, 'logout']);
+$routes->get('/filtro/productos/test', [CtrlApiProductos::class, 'index']);
```
Comment (CarlosJCD) [inline app/Config/Routes.php:77]: why call this test?

**Review by CarlosJCD — COMMENTED** [review]
Nice job @CarlosMay7, check out the comment i left you and remember to do the changes that were mentioned today about the button on the mobile views

**Review by FernandoJoachin — APPROVED** [review]

**Review by CarlosJCD — APPROVED** [review]

**Review by CheloChidoone — APPROVED** [review]
Good Job on the layout @CarlosMay7, @CarlosJCD  take note please that one detail is that some of the images on the list view appear to be too small and you can see a lot of empty space around it but maybe that fix can be added on the next PR of the filters so it doesn't delay anymore

File: app/Controllers/CtrlApiProductos.php [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+                ->findAll();
+        } else {
+            $dataResponse = (new ProductModel())
+                ->select('productName')
+                ->like('productName', "%{$data['productName']}%")
+                ->findAll();
+        }
+
+        return $this->response->setJSON($dataResponse);
+    }
```
Comment (LUGAMAFE) [inline app/Controllers/CtrlApiProductos.php:38]:
```suggestion
    public function getProductos()
    {
        $data = $this->getInputData();
        $query = $this->buildBaseQuery($data);

        if ($this->hasCategoryId($data)) {
            $query = $this->applyCategoryFilter($query, $data['categoryId']);
        }

        $dataResponse = $query->findAll();
        return $this->response->setJSON($dataResponse);
    }

    private function getInputData()
    {
        return json_decode($this->request->getBody(), true);
    }

    private function buildBaseQuery($data)
    {
        return (new ProductModel())
            ->select('productName')
            ->like('productName', "%{$data['productName']}%");
    }

    private function hasCategoryId($data)
    {
        return isset($data['categoryId']) && $data['categoryId'] !== '';
    }

    private function applyCategoryFilter($query, $categoryId)
    {
        return $query
            ->join('products-category_tags', 'products-category_tags.pctProductId = products.productId')
            ->where('productCategoryId', $categoryId)
            ->groupBy('productId');
    }
```

File: app/Controllers/CtrlApiProductos.php [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+use App\Models\ProductModel;
+
+class CtrlApiProductos extends BaseController
+{
+    public function index()
+    {
+        $categories = (new CategoryModel())->select('categoryId, categoryName')->findAll();
+
+        return view('public/filters', ['categories' => $categories]);
+    }
```
Comment (LUGAMAFE) [inline app/Controllers/CtrlApiProductos.php:15]: Are we using index() method?

File: app/Controllers/CtrlPublicPages.php [hunk 4]
```diff
@@ -55,17 +83,90 @@ public function viewEquipment(): string
 
     public function viewCategory(): string
     {
+        $filter        = $this->request->getGet();
         $categorySlug  = $this->request->getGet('categorias') ?? '';
         $categoryModel = new CategoryModel();
         $category      = $categoryModel->join('files', 'files.fileId = categories.categoryImageId')->where('categorySlug', $categorySlug)->first();
```
Comment (LUGAMAFE) [inline app/Controllers/CtrlPublicPages.php:89]: not used
```suggestion
```

File: app/Controllers/CtrlPublicPages.php [hunk 5]
```diff
@@ -55,17 +83,90 @@ public function viewEquipment(): string
...
+            'metaDescription'      => 'Explora los mejores productos de la categoría',
+            'category'             => $principalCategory,
+            'categories'           => $categories,
+            'categoryTags'         => $categoryTags,
+            'productTags'          => $productTags,
+            'products'             => $products,
+            'pagination'           => $pagination,
+            'clasificationOptions' => $clasificationAndOrderOptions,
+            'perPageOptions'       => $perPageOptions,
+        ]);
```
Comment (LUGAMAFE) [inline app/Controllers/CtrlPublicPages.php:169]: Can we refactor this code in a futher PR please to something like this:

```php
public function viewCategory(): string
{
    $filter = $this->initializeFilter($this->request->getGet());
    $categorySlug = $this->request->getGet('categorias') ?? '';
    $category = $this->getCategoryBySlug($categorySlug);

    $categories = $this->getCategories($filter['categorias']);
    $categoryTags = $this->getCategoryTags($filter);
    $productTags = $this->getProductTags($filter);

    $productBuilder = $this->initializeProductBuilder($filter);
    $pagination = $this->getPagination($filter, $productBuilder);

    $products = $this->getProducts($productBuilder, $filter, $pagination['page']);
    $principalCategory = $this->getPrincipalCategory($filter['categorias']);
    $clasificationAndOrderOptions = $this->getClasificationAndOrderOptions($filter);
    $perPageOptions = $this->getPerPageOptions($filter);

    return view('public/category', [
        'metaTitle'            => 'categorias',
        'metaDescription'      => 'Explora los mejores productos de la categoría',
        'category'             => $principalCategory,
        'categories'           => $categories,
        'categoryTags'         => $categoryTags,
        'productTags'          => $productTags,
        'products'             => $products,
        'pagination'           => $pagination,
        'clasificationOptions' => $clasificationAndOrderOptions,
        'perPageOptions'       => $perPageOptions,
    ]);
}
// Métodos privados para las tareas específicas (initializeFilter, getCategoryBySlug, getCategories, getCategoryTags, getProductTags, initializeProductBuilder, getPagination, getProducts, getPrincipalCategory, getClasificationAndOrderOptions, getPerPageOptions)
```

File: app/Models/ProductTagModel.php [hunk 6]
```diff
@@ -26,10 +25,20 @@ public function createProductTags(string $productId, array $productTags)
...
+            ->join('products', 'products.productId = product_tags.ptId')
+            ->join('categories', 'products.productCategoryId = categories.categoryId');
+
+        $productTags = (! empty($categorySlugs) ? $builder->whereIn('categories.categorySlug', $categorySlugs)->get() : $builder->get())->getResultArray();
+
+        return array_map(static function ($productTag) use ($productTagSlugs) {
+            $productTag['selected'] = in_array($productTag['ptSlug'], $productTagSlugs, true);
+
+            return $productTag;
+        }, $productTags);
```
Comment (LUGAMAFE) [inline app/Models/ProductTagModel.php:42]:
```suggestion
    public function getAllAndFormatSelected(array $productTagSlugs, array $categorySlugs)
    {
        $productTags = $this->getProductTagsFilteredByCategorySlugs($categorySlugs);
        return $this->formatProductTags($productTags, $productTagSlugs);
    }

    private function getProductTagsFilteredByCategorySlugs(array $categorySlugs)
    {
        $builder = $this->getProductTagsQueryBuilder();
        if (!empty($categorySlugs)) {
            $builder->whereIn('categories.categorySlug', $categorySlugs);
        }
        return $builder->get()->getResultArray();
    }

    private function getProductTagsQueryBuilder()
    {
        return $this->builder('product_tags')
            ->distinct()
            ->select(['ptName', 'ptSlug'])
            ->join('products', 'products.productId = product_tags.ptId')
            ->join('categories', 'products.productCategoryId = categories.categoryId');
    }

    private function formatProductTags(array $productTags, array $productTagSlugs)
    {
        return array_map(static function ($productTag) use ($productTagSlugs) {
            $productTag['selected'] = in_array($productTag['ptSlug'], $productTagSlugs, true);
            return $productTag;
        }, $productTags);
    }
```

**Review by LUGAMAFE — APPROVED** [review]
Porfavor chicos a pesar de que les puse correcciones, no es necesario que las hagan ahora. Refactorizemos el codigo en futuros PRs

**Pull request merged**